genie-agent/src/main/java/com/netflix/genie/agent/execution/process/impl/JobProcessManagerImpl.java

```diff
@@ -146,19 +148,43 @@ public void launchProcess(File jobDirectory,
     @Override
     public void kill(final KillService.KillSource source) {
         log.info("Killing job process (kill event source: {})", source);
-        // TODO: These may need to be done atomically as a tandem
         if (!this.killed.compareAndSet(false, true)) {
             // this job was already killed by something else
             return;
         }
         this.killSource.set(source);
-
         final Process process = this.processReference.get();
+        if (process == null) {
+            return;
+        }
 
-        // TODO: This probably isn't enough. We need retries with a force at the end
-        if (process != null) {
+        try {
+            // Grace killing period
+            final Instant graceKillEnd = Instant.now().plusSeconds(10);
```

Comment:
this hardcoded 10 seconds could be a place for improvement via property or at least should be a static constant as it is used multiple times and you'd probably want to change them all and not miss one just cause you didn't scan the code